velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window {
+
+namespace {
+
+class AggregateWindowFunction : public exec::WindowFunction {
+ public:
+  explicit AggregateWindowFunction(
+      const std::string& name,
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool) {
+    argTypes_.reserve(args.size());
+    argIndices_.reserve(args.size());
+    argVectors_.reserve(args.size());
+    for (const auto& arg : args) {
+      argTypes_.push_back(arg.type);
+      // TODO : Enhance code to handle constant arguments.
+      VELOX_CHECK_NULL(arg.constantValue);
+      VELOX_CHECK(arg.index.has_value());
+      argIndices_.push_back(arg.index.value());
+      argVectors_.push_back(BaseVector::create(arg.type, 0, pool_));
+    }
+    // Create an Aggregate function object to do result computation. Window
+    // function usage only requires single group aggregation for calculating
+    // the function value for each row.
+    aggregate_ = exec::Aggregate::create(
+        name, core::AggregationNode::Step::kSingle, argTypes_, resultType);
+
+    // Aggregate initialization.
+    int32_t rowSizeOffset = bits::nbytes(1);
+    singleGroupRowSize_ = rowSizeOffset + sizeof(int32_t);
+    int32_t nullOffset = 0;
+    aggregate_->setOffsets(
+        singleGroupRowSize_,
+        exec::RowContainer::nullByte(nullOffset),
+        exec::RowContainer::nullMask(nullOffset),
+        /* needed for out of line allocations */ rowSizeOffset);
+    singleGroupRowSize_ += aggregate_->accumulatorFixedWidthSize();
+
+    // Constructing the single row in the MemoryPool.
+    singleGroupRow_ = (char*)(pool_->allocate(singleGroupRowSize_));
```

Comment:
Use BufferPtr.

Reply:
The benefit of using BufferPtr is that it is a smart pointer, hence, memory free happens automatically. We should always prefer smart pointers to explicit allocation and free unless there is a very good reason not to).